Plutarch/TryFrom.hs

```diff
@@ -0,0 +1,400 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Plutarch.TryFrom (
+  PTryFrom (PTryFromExcess, ptryFrom),
+  PFrom (pfrom),
+  PMaybeFrom (PMaybeFromExcess, pmaybeFrom),
+) where
+
+import Data.Proxy (Proxy (Proxy))
+import GHC.TypeLits (KnownNat, Nat, natVal, type (+))
+
+import Plutarch.Builtin (
+  PAsData,
+  PBuiltinList,
+  PBuiltinMap,
+  PBuiltinPair,
+  PData,
+  PIsData (pfromData),
+  pasByteStr,
+  pasConstr,
+  pasInt,
+  pasList,
+  pasMap,
+  pdata,
+  pforgetData,
+  pfstBuiltin,
+  ppairDataBuiltin,
+  psndBuiltin,
+ )
+
+import Plutarch.Maybe (PMaybe)
+
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal.Other (
+  POpaque,
+  PType,
+  Term,
+  pcon,
+  perror,
+  plam,
+  plet,
+  (#),
+  (#$),
+  type (:-->),
+ )
+import Plutarch.Unit (PUnit (PUnit))
+
+import Plutarch.DataRepr.Internal (
+  PDataRecord,
+  PDataSum,
+  PLabeledType ((:=)),
+  pdcons,
+  pdnil,
+ )
+
+import Plutarch.Bool (pif, (#==))
+
+import Plutarch.List (
+  phead,
+  pmap,
+  ptail,
+ )
+
+import Plutarch.Unsafe (punsafeBuiltin, punsafeCoerce)
+import qualified PlutusCore as PLC
+
+import Plutarch.TermCont (TermCont (TermCont, runTermCont), tcont, unTermCont)
+
+----------------------- The class PTryFrom ----------------------------------------------
+
+{- |
+    This checks the datastructure for validity.
+    If you don't care about parts of the datastructure
+    don't verify those parts, just let it return a PData
+    instead
+    Be aware this might get really expensive, so only
+    use it if you cannot establish trust otherwise
+    (e.g. via only checking a part of your Data with
+    PTryFrom)
+    Laws:
+      - the operation `ptryFrom` mustn't change the representation of the underlying data
+      - the operation `ptryFrom` must always prove the integrity of the whole target type
+        - example:
+          `ptryFrom PData (PAsData (PBuiltinList PData))` must only succeed if the underlying
+          representation is a `BuiltinList` containing any `PData`
+        - all conversion are fallible, this happens if the representation doesn't match
+          the expected type.
+      - the operation `ptryFrom` proves equality between the less expressive `PType` `a` and
+        the more expressive `PType` `b`, hence the first element of the resulting Tuple
+        must always be wrapped in `PAsData` if the origin type was `PData` (see law 1)
+      - the result type `b` must always be safe than the origin type `a`, i.e. it must carry
+        more information
+-}
+class PTryFrom (a :: PType) (b :: PType) where
+  type PTryFromExcess a b :: PType
```

Comment:
I think the issue I have is that rather than returning this from `ptryFrom`, we really want to somehow splice in some code into the verification process, such that we're traversing while verifying essentially.

Reply:
You can emulate this for simple types by having `PTryFromExcess` be a list of variables referring to all the intermediately generated terms (this isn't trivially possible, but let's ignore the irrelevant difficulties for now). Maybe rather than having users do `ptryFrom` on complex types, we should somehow expose types that aren't verified yet. Let me think more about this.